lightning/src/ln/channelmanager.rs

```diff
@@ -291,6 +291,32 @@ struct ReceiveError {
 	msg: &'static str,
 }
 
+/// This enum is used to specify which error data to send to peers when failing back an HTLC
+/// using [`ChannelManager::fail_htlc_backwards_with_reason`].
+///
+/// For more info on failure codes, see <https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages>.
+#[derive(Clone, Copy)]
+pub enum FailureCode {
+	/// A general temporary failure for otherwise unspecified errors that occur
+	/// for the entire node while handling the received onion. This can be used in situations
+	/// where other temporary error codes would not fit, while indicating to the origin node
+	/// it may want to retry the payment.
+	TemporaryNodeFailure             = 0x2000 | 2,
+	/// We have a required feature which was not in this onion.
+	/// You may wish to use this in cases where you require some additional metadata be provided
+	/// with the payment but which was not provided here.
+	RequiredNodeFeatureMissing       = 0x4000 | 0x2000 | 3,
+	/// The CLTV expiry is too close to the current block height for safe handling by us. This
+	/// should only be used when failing back as a forwarding node.
```

Comment:
Note that `fail_with_reason` is only ever called in response to a `PaymentClaimable` event, so it'll never be called by a forwarding node 

```suggestion
	/// The CLTV expiry is too close to the current block height for safe handling. Useful if 
	/// the payer had an outdated view of your CLTV expiry requirements on send.
```
I'm not sure if this is exactly the right reason, but it would be good to provide an example of some use case for this error code

Reply:
> Note that fail_with_reason is only ever called in response to a PaymentClaimable event, so it'll never be called by a forwarding node

Ah I was thinking that, but wasn't sure if it might've been able to be used in a forwarding context, but that makes sense.

With that, in the spec it seems that final nodes are not supposed to specify `expiry_too_soon` when returning an error because differentiating this failure code (along with `incorrect_payment_amount`) allows for probing for the destination--for final nodes it merges all these failure codes into `incorrect_or_unknown_payment_details`. I'm wondering should this failure code be excluded then?

Reply:
Ah that's a good point.. @TheBlueMatt can confirm, I think we shouldn't use failure codes that are only intended to be used by forwarding nodes?

Reply:
Makes sense, I'll remove this variant (sometime tomorrow morning)